src/core/dbus-execute.c

```diff
@@ -1701,22 +1709,31 @@ int bus_exec_context_set_transient_property(
 
                 return 1;
 
-        } else if (streq(name, "RuntimeDirectoryMode")) {
+        } else if (STR_IN_SET(name, "RuntimeDirectoryMode", "StateDirectoryMode", "CacheDirectoryMode", "LogsDirectoryMode", "ConfigurationDirectoryMode")) {
                 mode_t m;
 
                 r = sd_bus_message_read(message, "u", &m);
                 if (r < 0)
                         return r;
 
                 if (mode != UNIT_CHECK) {
-                        c->runtime_directory_mode = m;
-
-                        unit_write_drop_in_private_format(u, mode, name, "RuntimeDirectoryMode=%040o", m);
+                        if (streq(name, "RuntimeDirectoryMode"))
+                                c->directories[EXEC_DIRECTORY_RUNTIME].mode = m;
+                        else if (streq(name, "StateDirectoryMode"))
+                                c->directories[EXEC_DIRECTORY_STATE].mode = m;
+                        else if (streq(name, "CacheDirectoryMode"))
+                                c->directories[EXEC_DIRECTORY_CACHE].mode = m;
+                        else if (streq(name, "LogsDirectoryMode"))
+                                c->directories[EXEC_DIRECTORY_LOGS].mode = m;
+                        else /* ConfigurationDirectoryMode */
+                                c->directories[EXEC_DIRECTORY_CONF].mode = m;
```

Comment:
I'd think coding this like this would be nicer:

```
static const char * const names[_EXEC_DIRECTORY_MAX] = {
        [EXEC_DIRECTORY_RUNTIME] = "RuntimeDirectoryMode",
        [EXEC_DIRECTORY_STATE] = "StateDirectoryMode",  
        …
};

for (i = 0; i < _EXEC_DIRECTORY_MAX; i++) 
        if (streq(name, names[i])) {
                c->directories[i] = m;
                break;
        }
```

i.e. always make use of the fact that this is an enum now...